include/sapi/implementation.h

```diff
@@ -1,439 +0,0 @@
-#define  RSA_KEY_SIZE_BITS_1024     RSA_ALLOWED_KEY_SIZE_1024
-#define  RSA_KEY_SIZE_BITS_2048     RSA_ALLOWED_KEY_SIZE_2048
-#define  MAX_RSA_KEY_BITS           2048
-#define  MAX_RSA_KEY_BYTES          256
-#define  AES_KEY_SIZES_BITS         {128,256}
-#define  AES_KEY_SIZE_BITS_128      AES_ALLOWED_KEY_SIZE_128
-#define  AES_KEY_SIZE_BITS_256      AES_ALLOWED_KEY_SIZE_256
-#define  MAX_AES_KEY_BITS           256
-#define  MAX_AES_KEY_BYTES          32
-#define MAX_AES_BLOCK_SIZE_BYTES				      \
-    MAX(AES_128_BLOCK_SIZE_BYTES,					\
-	MAX(AES_256_BLOCK_SIZE_BYTES, 0))
-#define  SM4_KEY_SIZES_BITS         {128}
-#define  SM4_KEY_SIZE_BITS_128      SM4_ALLOWED_KEY_SIZE_128
-#define  MAX_SM4_KEY_BITS           128
-#define  MAX_SM4_KEY_BYTES          16
-#define MAX_SM4_BLOCK_SIZE_BYTES			\
-    MAX(SM4_128_BLOCK_SIZE_BYTES, 0)
-#define  CAMELLIA_KEY_SIZES_BITS    {128}
-#define  CAMELLIA_KEY_SIZE_BITS_128    CAMELLIA_ALLOWED_KEY_SIZE_128
-#define  MAX_CAMELLIA_KEY_BITS      128
-#define  MAX_CAMELLIA_KEY_BYTES     16
-#define MAX_CAMELLIA_BLOCK_SIZE_BYTES				\
-    MAX(CAMELLIA_128_BLOCK_SIZE_BYTES, 0)
-
-// From Vendor-Specific: Table 5 - Defines for Implemented Curves
-
-#define  ECC_NIST_P256         YES
-#define  ECC_NIST_P384         YES
-#define  ECC_BN_P256           YES
-#define  ECC_CURVES							\
-    {TPM_ECC_BN_P256, TPM_ECC_NIST_P256, TPM_ECC_NIST_P384}
-#define  ECC_KEY_SIZES_BITS    {256, 384}
-#define  ECC_KEY_SIZE_BITS_256
-#define  ECC_KEY_SIZE_BITS_384
-#define  MAX_ECC_KEY_BITS      384
-#define  MAX_ECC_KEY_BYTES     48
-
-// From Vendor-Specific: Table 7 - Defines for Implementation Values
-
-#define FIELD_UPGRADE_IMPLEMENTED	NO	/* temporary define */
-#define BSIZE				UINT16	/* size used for internal storage of the size field of a TPM2B */
-#define BUFFER_ALIGNMENT		4	/* sets the size granularity for the buffers in a TPM2B structure */
-#define IMPLEMENTATION_PCR		24	/* the number of PCR in the TPM */
-#define PLATFORM_PCR			24	/* the number of PCR required by the relevant platform specification */
-#define DRTM_PCR			17	/* the DRTM PCR */
-#define HCRTM_PCR			0	/* PCR that will receive the H-CRTM value at
-						   TPM2_Startup.  This value should not be changed. */
-#define NUM_LOCALITIES			5	/* the number of localities supported by the TPM */
-#define MAX_HANDLE_NUM			3	/* the maximum number of handles in the handle area */
-#define MAX_ACTIVE_SESSIONS		64	/* the number of simultaneously active sessions that
-						   are supported by the TPM implementation */
-#define CONTEXT_SLOT                     UINT16	/* the type of an entry in the array of
-						   saved contexts */
-#define CONTEXT_COUNTER                  UINT64	/* the type of the saved session counter */
-#define MAX_LOADED_SESSIONS		3	/* the number of sessions that the TPM may have in memory */
-#define MAX_SESSION_NUM 		3	/* this is the current maximum value */
-#define MAX_LOADED_OBJECTS		3	/* the number of simultaneously loaded objects that
```

Comment:
Maybe it will be easier to split this PR into two or even three separate PRs, first including 57dc144, 200dd20, and ae3ea57, second just f5a843f (TPM -> TPM2) and the last 93d8eba, which removes the implementation.h? That way we can take the first three, and then work on the rest. I don't want to push my SYS_CONTEXT before this one, but I'm a bit blocked.

Reply:
I'm guessing these were all bundled together to minimize thrash in the stuff that we expose to clients. Every change here could mean a code change for a downstream consumer so we should try to minimize that stuff. Adding back at least the macros that the tabrmd and tools use shouldn't be too bad. Any in interest in working on the tools while I'm doing the tabrmd?

Reply:
Yep I can make the tools changes.